driver-embedded/src/main/com/mongodb/embedded/client/EmbeddedInternalConnection.java

```diff
@@ -0,0 +1,226 @@
+/*
+ * Copyright 2008-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mongodb.embedded.client;
+
+import com.mongodb.MongoCompressor;
+import com.mongodb.ServerAddress;
+import com.mongodb.async.SingleResultCallback;
+import com.mongodb.connection.AsyncCompletionHandler;
+import com.mongodb.connection.ClusterId;
+import com.mongodb.connection.ConnectionDescription;
+import com.mongodb.connection.ServerId;
+import com.mongodb.connection.Stream;
+import com.mongodb.connection.StreamFactory;
+import com.mongodb.event.CommandListener;
+import com.mongodb.internal.connection.Authenticator;
+import com.mongodb.internal.connection.CommandMessage;
+import com.mongodb.internal.connection.InternalConnection;
+import com.mongodb.internal.connection.InternalStreamConnection;
+import com.mongodb.internal.connection.InternalStreamConnectionInitializer;
+import com.mongodb.internal.connection.ResponseBuffers;
+import com.mongodb.session.SessionContext;
+import com.sun.jna.Pointer;
+import com.sun.jna.ptr.IntByReference;
+import com.sun.jna.ptr.PointerByReference;
+import org.bson.BsonDocument;
+import org.bson.ByteBuf;
+import org.bson.ByteBufNIO;
+import org.bson.codecs.Decoder;
+
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+class EmbeddedInternalConnection implements InternalConnection {
+    private final InternalConnection wrapped;
+    private volatile Pointer clientPointer;
+
+    EmbeddedInternalConnection(final Pointer databasePointer, final CommandListener commandListener,
+                               final BsonDocument clientMetadataDocument) {
+        this.clientPointer = MongoDBCAPIHelper.db_client_new(databasePointer);
```

Comment:
The connection is responsible for the expensive creation of the clientPointer. This is a simple wrapping of the previously used internal connection. However, combining with a connection pool allows this connection to be reused.